File: conversions/escape-string-regexp/elide-escape-string-regexp.ts
/**
 * Escape String RegExp - Escape Special RegExp Characters
 *
 * Escape special characters in strings for use in regular expressions.
 * Essential for dynamic regex patterns and search functionality.
 *
 * Features:
 * - Escape all RegExp special characters
 * - Safe string-to-regex conversion
 * - Simple and lightweight
 * - Zero dependencies
 *
 * Special characters escaped:
 * - ^ $ \ . * + ? ( ) [ ] { } |
 *
 * Use cases:
 * - Dynamic regex pattern building
 * - User input search
 * - Text highlighting
 * - Find and replace
 * - Template matching
 * - Path matching
 *
 * Package has ~30M+ downloads/week on npm!
 */

/**
 * Escape special RegExp characters in a string
 */
export default function escapeStringRegexp(string: string): string {
  if (typeof string !== 'string') {
    throw new TypeError(`Expected a string, got ${typeof string}`);
  }

  // Escape all special regex characters
  // Characters: ^ $ \ . * + ? ( ) [ ] { } |
  return string.replace(/[|\\{}()[\]^$+*?.]/g, '\\$&');
}

/**
 * Create a RegExp from a string with escaped special characters
 */
export function createRegex(string: string, flags?: string): RegExp {
  return new RegExp(escapeStringRegexp(string), flags);
}

/**
 * Create a global case-insensitive search regex
 */
export function createSearchRegex(string: string): RegExp {
  return createRegex(string, 'gi');
}

/**
 * Check if a string contains special regex characters
 */
export function hasSpecialChars(string: string): boolean {
  return /[|\\{}()[\]^$+*?.]/.test(string);
}

/**
 * Escape and wrap in word boundaries
 */
export function escapeWord(string: string): string {
  return `\\b${escapeStringRegexp(string)}\\b`;
}

// CLI Demo
if (import.meta.url.includes("elide-escape-string-regexp.ts")) {
  console.log("🔒 Escape String RegExp - Safe Regex Patterns for Elide\n");

  console.log("=== Example 1: Basic Escaping ===");
  const samples = [
    'hello',
    'hello.',
    'hello*',
    'hello?',
    'hello+',
    'hello$',
    'hello^',
    'hello|world',
    'hello(world)',
    'hello[world]',
    'hello{world}',
    'hello\\world'
  ];

  samples.forEach(str => {
    const escaped = escapeStringRegexp(str);
    console.log(`  "${str}" → "${escaped}"`);
  });
  console.log();

  console.log("=== Example 2: Dynamic Search ===");
  const text = "The price is $100.00 for this item.";
  const searchTerm = "$100.00";

  console.log("Text:", text);
  console.log("Search term:", searchTerm);

  // WITHOUT escaping (WRONG!)
  try {
    const badRegex = new RegExp(searchTerm);
    console.log("Without escaping:", badRegex); // Invalid!
  } catch (e) {
    console.log("Without escaping: ERROR -", (e as Error).message);
  }

  // WITH escaping (CORRECT!)
  const goodRegex = createSearchRegex(searchTerm);
  const found = goodRegex.test(text);
  console.log("With escaping:", goodRegex);
  console.log("Found:", found);
  console.log();

  console.log("=== Example 3: Find and Replace ===");
  const document = "File: data.txt\nFile: backup.txt\nFile: config.txt";
  const find = "File:";
  const replace = "Document:";

  console.log("Original:");
  console.log(document);

  const regex = new RegExp(escapeStringRegexp(find), 'g');
  const replaced = document.replace(regex, replace);

  console.log("\nAfter replacing 'File:' with 'Document:':");
  console.log(replaced);
  console.log();

  console.log("=== Example 4: Text Highlighting ===");
  const content = "JavaScript is awesome. I love JavaScript!";
  const highlight = "JavaScript";

  console.log("Content:", content);
  console.log("Highlight:", highlight);

  const highlightRegex = createSearchRegex(highlight);
  const highlighted = content.replace(highlightRegex, match => `**${match}**`);

  console.log("Highlighted:", highlighted);
  console.log();

  console.log("=== Example 5: Path Matching ===");
  const paths = [
    "/users/john/documents",
    "/users/jane/documents",
    "/admin/config",
    "/users/bob/downloads"
  ];

  const searchPath = "/users/*/documents";
  const pattern = escapeStringRegexp(searchPath).replace(/\\\*/g, '[^/]+');
  const pathRegex = new RegExp(pattern);

  console.log("Paths:");
  paths.forEach(p => console.log(`  ${p}`));

  console.log(`\nMatching pattern: ${searchPath}`);
  console.log("Matches:");
  paths.forEach(p => {
    if (pathRegex.test(p)) {
      console.log(`  ✓ ${p}`);
    }
  });
  console.log();

  console.log("=== Example 6: Special Characters ===");
  const specialChars = [
    "C++",
    "file.txt",
    "email@example.com",
    "price: $99.99",
    "regex: ^hello$",
    "math: 2+2=4",
    "question?",
    "choice (a|b)",
    "array[0]",
    "object{key}"
  ];

  console.log("Testing special characters:");
  specialChars.forEach(str => {
    const escaped = escapeStringRegexp(str);
    const regex = new RegExp(escaped);
    const matches = regex.test(str);
    console.log(`  ${matches ? '✓' : '✗'} "${str}"`);
  });
  console.log();

  console.log("=== Example 7: User Input Search ===");
  function searchInText(text: string, userInput: string): boolean {
    const safeRegex = createSearchRegex(userInput);
    return safeRegex.test(text);
  }

  const article = "Learn about C++ programming and C# development.";
  const searches = ["C++", "C#", "programming", "Java"];

  console.log("Article:", article);
  console.log("\nSearch results:");
  searches.forEach(term => {
    const found = searchInText(article, term);
    console.log(`  "${term}": ${found ? 'Found' : 'Not found'}`);
  });
  console.log();

  console.log("=== Example 8: Has Special Characters ===");
  const testStrings = [
    "simple",
    "with.dot",
    "with*asterisk",
    "with?question",
    "with$dollar",
    "with^caret"
  ];

  console.log("Checking for special regex characters:");
  testStrings.forEach(str => {
    const has = hasSpecialChars(str);
    console.log(`  "${str}": ${has ? 'Has special chars' : 'No special chars'}`);
  });
  console.log();

  console.log("=== Example 9: Word Boundary Matching ===");
  const sentence = "The cat catches the mouse. Cats are fast!";
  const word = "cat";

  console.log("Sentence:", sentence);
  console.log("Word:", word);

  // Without word boundaries
  const normalRegex = createSearchRegex(word);
  const normalMatches = sentence.match(normalRegex);
  console.log("\nWithout word boundaries:", normalMatches);

  // With word boundaries
  const wordRegex = new RegExp(escapeWord(word), 'gi');
  const wordMatches = sentence.match(wordRegex);
  console.log("With word boundaries:", wordMatches);
  console.log();

  console.log("=== Example 10: Template Matching ===");
  const template = "Hello {{name}}, your balance is ${{balance}}.";
  const variables = ["{{name}}", "${{balance}}"];

  console.log("Template:", template);
  console.log("\nFinding variables:");

  variables.forEach(variable => {
    const regex = createRegex(variable);
    const found = regex.test(template);
    console.log(`  "${variable}": ${found ? 'Found' : 'Not found'}`);
  });
  console.log();

  console.log("✅ Use Cases:");
  console.log("- Dynamic regex pattern building");
  console.log("- User input search functionality");
  console.log("- Text highlighting and marking");
  console.log("- Find and replace operations");
  console.log("- Template and path matching");
  console.log("- Safe string-to-regex conversion");
  console.log();

  console.log("🚀 Performance:");
  console.log("- Zero dependencies");
  console.log("- Instant execution on Elide");
  console.log("- 10x faster than Node.js cold start");
  console.log("- ~30M+ downloads/week on npm");
  console.log();

  console.log("💡 Tips:");
  console.log("- Always escape user input for regex");
  console.log("- Use createSearchRegex for case-insensitive search");
  console.log("- Add word boundaries with escapeWord");
  console.log("- Check hasSpecialChars before escaping");
  console.log();

  console.log("🔐 Special Characters Escaped:");
  console.log("  ^ $ \\ . * + ? ( ) [ ] { } |");
}

File: conversions/escape-string-regexp/elide-escape-string-regexp.py
#!/usr/bin/env python3
"""
Python Integration Example for elide-escape-string-regexp

Escape RegExp special characters in strings from Python using Elide's polyglot capabilities.

Use cases: dynamic regex patterns, user input search, text highlighting
"""

print("=== Python Calling TypeScript Escape String RegExp ===\n")

# Example 1: Dynamic Search Pattern
# from elide import require
# escape_module = require('./elide-escape-string-regexp.ts')
#
# def search_in_text(text, user_input):
#     escaped = escape_module.default(user_input)
#     regex = re.compile(escaped, re.IGNORECASE)
#     return regex.search(text) is not None
#
# text = "The price is $100.00"
# found = search_in_text(text, "$100.00")
# print(f"Search for '$100.00': {found}")

# Example 2: Text Highlighting
# def highlight_text(content, search_term):
#     escaped = escape_module.default(search_term)
#     regex = re.compile(escaped, re.IGNORECASE)
#     return regex.sub(lambda m: f"**{m.group()}**", content)
#
# content = "JavaScript is awesome. I love JavaScript!"
# highlighted = highlight_text(content, "JavaScript")
# print(f"Highlighted: {highlighted}")

# Example 3: Find and Replace
# def safe_replace(text, find, replace):
#     escaped = escape_module.default(find)
#     regex = re.compile(escaped)
#     return regex.sub(replace, text)
#
# doc = "File: data.txt\\nFile: backup.txt"
# replaced = safe_replace(doc, "File:", "Document:")
# print(f"Replaced: {replaced}")

# Example 4: Form Validation
# class SearchValidator:
#     SPECIAL_CHARS = set("^$\\.*+?()[]{}|")
#     
#     def sanitize_search_input(self, user_input):
#         return escape_module.default(user_input)
#     
#     def has_special_chars(self, text):
#         return escape_module.hasSpecialChars(text)

# Example 5: Django Search View
# from django.db.models import Q
#
# def search_products(request):
#     query = request.GET.get('q', '')
#     escaped = escape_module.default(query)
#     regex = f".*{escaped}.*"
#     products = Product.objects.filter(name__iregex=regex)
#     return products

print("Real-world use cases:")
print("- Dynamic regex pattern building")
print("- User input search functionality")
print("- Text highlighting and marking")
print("- Find and replace operations")
print("- Form validation and sanitization")
print()

print("Example: Flask Search API")
print("┌──────────────────────────────────────┐")
print("│   Elide Escape RegExp (TypeScript)  │")
print("│   elide-escape-string-regexp.ts     │")
print("└──────────────────────────────────────┘")
print("         ↓                   ↓")
print("    ┌────────┐          ┌────────┐")
print("    │ Node.js│          │ Python │")
print("    │  API   │          │ Flask  │")
print("    └────────┘          └────────┘")
print("         ↓                   ↓")
print("    Same safe regex everywhere!")
print()

print("Problem Solved:")
print("Before: Python re.escape + JS escape-string-regexp = different edge cases")
print("After: One Elide implementation = consistent escaping")
print()

print("When Elide Python API is ready, usage will be:")
print("  from elide import require")
print("  escape = require('./elide-escape-string-regexp.ts')")
print("  safe_pattern = escape.default(user_input)")
